Automatically update internal links

david-venhoff opened the pull request:
### Short description
<!-- Describe this PR in one or two sentences. -->
This pr adds a feature to automatically update internal links when the content object they link to gets updated.
This also works across regions.
Right now no address is shown as link text for links to pois as requested in #1878, is that still wanted (And if so, do you have a preferred format @svenseeberg)?
Also I still need to test if the performance is okay. Ideally I would like to do that on the test system, as the data are more realistic there.


### Proposed changes
<!-- Describe this PR in more detail. --> 
- When saving a page, fix all links in that page that are outdated
- When saving a page, find all links that point to this page, and for each link update the content translation of that link
- Add the attribute `data-integreat-auto-update` and automatically create a link text for those links if possible
- Add icons before event and poi links


### Side effects
<!-- List all related components that have not been explicitly changed but may be affected by this PR -->
- When a user updates a page and if that causes another page to be updated, the author will be shown as the user who changed the other page, even if the user does not have permission to do so (Should alternatively no user be shown?)


### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1878


__________________________________________________
<!-- Keep this link for the potential reviewer -->
[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

seluianova commented:
> Add icons before event and poi links
Should it be the default icon or a specific location/event icon?

david-venhoff commented:
> Should it be the default icon or an icon set for the particular location/event ?

Seems like a good idea to me

> Is it okay, that we add this attribute to external links as well?

While technically redundant, they are ignored internally, so imo that is not a problem

seluianova commented:
> While technically redundant, they are ignored internally, so imo that is not a problem

I'm not insisting here, but perhaps it can be understood to mean that we will update the link text for external links as well

david-venhoff commented:
I have added the functionality to load custom defined icons, which looks like this:
![grafik](https://github.com/digitalfabrik/integreat-cms/assets/78504586/97f9a633-7cfe-4510-be7e-45116e4aa3f7)

I have also adjusted the behavior of the link dialog, so that either the text and url input can be edited, or that the 'link to internal content' input and the checkbox can be edited. Unfortunately, tinymce is very buggy and does not visually display which inputs are enabled and which are disabled. I also tried disable the checkbox on initial load, but that causes another tinymce bug where it will stay visually disabled, even when it gets enabled later…
I feel like the current custom link input dialog is hard to maintain and the tinymce api is very limiting. Maybe we should create an issue to completely refactor it at some point? 

> 4. Is it okay that the link text with an icon will be shown like this?

I think yes, because it accurately represents what the link looks like and users can expand the text field to view the full link text. If you don't agree, would you prefer just  `[image]title` for links with both image and text?

sarahsporck commented on integreat_cms/cms/utils/internal_link_utils.py:
```diff
@@ -51,14 +60,14 @@ def update_link_language(
-                link_text = fixed_link
+        # Update the title if it was previously the url, otherwise use the new title
+        link_html = None
+        if len(link) == 0 and link.text and current_url.strip() == link.text.strip():
```
What does the len(link) check for? Isn't the link an Element?

seluianova replied:
`# Update the title if it was previously the url `

I haven't looked into it deeply, but it seems to work for pages, but not for events or locations 🤔

sarahsporck commented on integreat_cms/cms/views/utils/search_content_ajax.py:
```diff
@@ -44,9 +45,19 @@ def format_object_translation(
     :param typ: The type of this object
     :return: A dictionary with the title, path, url and type of the translation object
     """
+    if isinstance(object_translation.link_title, str):
```
What is the difference between `object_translation.link_title` and `object_translation.title`?

david-venhoff replied:
`link_title` is the method I defined in `integreat_cms/cms/models/abstract_content_translation.py` which can both return a `str` and an `Element`. I did this because the issue for this pr wants icons displayed next to the links for internal content translations. So in those cases, `link_title` returns the icon as an HTML Element in addition to the normal title

sarahsporck commented on integreat_cms/static/src/js/tinymce-plugins/custom_link_input/plugin.js:
```diff
@@ -143,6 +144,16 @@ import { getCsrfToken } from "../../utils/csrf-token";
                     api.disable("submit");
                 }
 
+                if (data.completions !== "") {
+                    api.disable("text");
+                    api.disable("url");
+                    api.enable("autoupdate");
+                } else {
+                    api.enable("text");
+                    api.enable("url");
+                    api.disable("autoupdate");
+                }
+
```
So...what's this doing? And why? :)

david-venhoff replied:
This enables the checkbox to automatically update an internal link only when there is a current completion, and it disables the manual inputs for a link, too in that case. Unfortunately it does not display correctly, due to a bug in tinymce

sarahsporck commented on integreat_cms/cms/utils/content_translation_utils.py:
```diff
@@ -0,0 +1,170 @@
+            outdated_content_translation.content,
+            outdated_content_translation.language.slug,
+        )
+        if new_content == outdated_content_translation.content:
+            continue
```
What's this check for?

david-venhoff replied:
This is just an optimization so that no new translations get created when the content did not change. For example, this function gets executed when the status of a content translation changes to `PUBLIC`, but its title and content might not have changed and all existing links are still valid. In this case there would be no point in saving new versions with the same content.

seluianova reviewed (commented):
> Unfortunately, tinymce is very buggy and does not visually display which inputs are enabled and which are disabled

To be honest I think it was better when the user could control the checkbox regardless of the link type.

Because now you can open a dialog, can _uncheck_ the checkbox, but can't _check_ it back. 
Or, if you edit a link, where the checkbox was unchecked, you can _check_ it. But then you can't _uncheck_ it back!
So it's like - you still can edit a checkbox, but only once 😃
I think it's just more confusing and looks like a bug 🙈

MariaKabanova commented:
I think it’s an awesome change which adds a lot of value for customers. I’ve tested this pull request, and I found a couple of small things.

The first one is about an icon. Looks like even though the link title and url are auto-updated, the icon of the internal link is not updated automatically. You can follow the next steps to reproduce that: 

1. Open page, then insert the link on another page
2. Go to the linked page and change or delete an icon there
3. Return to the original page, where you can see that icon didn’t change. But if you click on “update”, the icon is changed after all

The second one contains the similar problem but in imprint. After inserting a link to the page in imprint, this link doesn’t automatically update. Steps to reproduce:

1. Open Imprint, then create the link to another page (insert/link)
2. Go to the linked page and change its title
3. Return to the imprint and see that the title didn’t change
4. Click on “update” and see that title has changed

I'm not sure it's worth fixing, please let me know what you think.

MariaKabanova commented:
@david-venhoff  

Today I've tested again this branch and now everything is great!  It means that  now we need to wait for #1627 is fixed. After that I'll retest imprints. :-)

seluianova commented on integreat_cms/cms/models/events/event_translation.py:
```diff
@@ -83,6 +85,24 @@ def backend_edit_link(self) -> str:
+        if icon := self.event.icon:
+            if thumbnail := icon.thumbnail_url:
+                img = Element("img")
+                img.set("style", "height:15px;")
```
```suggestion
                img.set("style", "height:15px; margin-right:5px;")
```

david-venhoff replied:
I also added some left margin for rtl languages

seluianova commented on integreat_cms/cms/models/pages/page_translation.py:
```diff
@@ -385,6 +386,23 @@ def path(self) -> SafeString:
+        if icon := self.page.icon:
+            if thumbnail := icon.thumbnail_url:
+                img = Element("img")
+                img.set("style", "height:15px;")
```
```suggestion
                img.set("style", "height:15px; margin-right:5px;")
```

seluianova commented on integreat_cms/cms/models/pois/poi_translation.py:
```diff
@@ -94,6 +96,24 @@ def backend_edit_link(self) -> str:
+        if icon := self.poi.icon:
+            if thumbnail := icon.thumbnail_url:
+                img = Element("img")
+                img.set("style", "height:15px;")
```
```suggestion
                img.set("style", "height:15px; margin-right:5px;")
```

seluianova reviewed (approved):
LGTM, thanks!

**UPD**. tooo minor and may be fixed later or not fixed at all, but as a note:
If you update only an icon of the page/event/location, it's not updated automatically (as it works with titles), but only if you click "Update" on the page, where the corresponding link is used

seluianova commented on integreat_cms/cms/models/events/event_translation.py:
```diff
@@ -83,6 +85,24 @@ def backend_edit_link(self) -> str:
         )
 
+    @cached_property
+    def link_title(self) -> Element | str:
```
And I think we could do some refactoring here to reduce code duplication.
The implementation is almost the same for all content types. I think we could basically keep one default method in abstract_content_translation.py and only re-assign the default icon (null for pages, not-null for events and locations)

david-venhoff commented:
> If you update only an icon of the page/event/location, it's not updated automatically (as it works with titles), but only if you click "Update" on the page, where the corresponding link is used

Oh, right. Unfortunately this is a little bit more complicated to fix, because the icon is not part of the translation forms. I'd prefer to fix this in a follow-up pr, if that is okay.

david-venhoff commented:
> If you update only an icon of the page/event/location, it's not updated automatically (as it works with titles), but only if you click "Update" on the page, where the corresponding link is used

That should already be implement (in 98b3c47fe5d87aa359e2f2f5fa179b9962ddd8b1)

MizukiTemma reviewed (approved):
Cool feature 😻 Thank you so much for tackling this issue!

As #1627 is in the backlog and therfore will not be implemented soon, I suggest to merge this PR already 👍

david-venhoff commented:
Quick status update: I would like to do some performance testing before merging this pr, so I'll block it on #2516

david-venhoff commented:
I tested this with production data and, as expected, the performance was really bad. I am currently working on increasing the performance, so that this pr can be merged.

david-venhoff commented:
@seluianova @MizukiTemma I have added an optimization to reduce the time it takes to save. On the production data saving a page now takes one or two seconds longer than without this pr, which I think is acceptable. Could you take a look at the changes?

MizukiTemma commented:
> @seluianova @MizukiTemma I have added an optimization to reduce the time it takes to save. On the production data saving a page now takes one or two seconds longer than without this pr, which I think is acceptable. Could you take a look at the changes?

Sorry for having overlooked this request 🙈 Could you rebase and resolve the conflicts? Then I will check this PR again 💪

david-venhoff commented:
The pr is now rebased onto develop :)

MizukiTemma commented:
@david-venhoff Thank you for rebasing 👍 
I found only one thins to fix. `Cannot resolve keyword 'slug' into field` occurs when a link is added to the imprint.
(Sorry, this is an extra work due to my recent PR adding Imprint to the link check 🙏 )

david-venhoff commented:
Great catch!
